### ISSUE TITLE
Reconcilers should rely on knative/pkg to Update TaskRuns and PipelineRuns

### ISSUE THREAD

**Issue by imjasonh** [tektoncd/pipeline#5146]
# Background

The TaskRun and PipelineRun controllers are invoked by Knative controller code that calls `ReconcileKind(ctx, tr)` with the result of a K8s watch on resources, then determines if those resources need to be updated, and if so, calls `Update` on them.

This is good, because Knative code handles the watching and updating for us, and only has to assume that any changes Tekton wants to make to resources is expressed in terms of updates to the `tr` object passed to `ReconcileKind`.

knative/pkg expects this:

```
knative/pkg calls Watch
  -> ReconcileKind(tr)
    -> if tr has changed, knative/pkg calls Update
```

However, our `ReconcileKind`s for TaskRuns and PipelineRuns (notably, not Runs) also make their own calls `Update`, as part of `updateLabelsAndAnnotations`:

https://github.com/tektoncd/pipeline/blob/8a7b0cfa755038f4cfdcd88c314a72a90bcab1a2/pkg/reconciler/pipelinerun/pipelinerun.go#L1171-L1186

https://github.com/tektoncd/pipeline/blob/8a7b0cfa755038f4cfdcd88c314a72a90bcab1a2/pkg/reconciler/taskrun/taskrun.go#L548-L573

Effectively, what we're doing is:

```
knative/pkg calls Watch
  -> ReconcileKind(pr)
    -> if pr has changed, Tekton calls Update
      -> if pr has changed (unlikely), knative/pkg calls Update
```

This leads to an unnecessary and unexpected change in the control of the update lifecycle for both PipelineRuns and TaskRuns. A method unsuspiciously called `updateLabelsAndAnnotations` is actually responsible for persisting all the changes we make during a reconciliation.

This isn't a _problem_ per se, it's just unexpected, and not how Knative reconciler code expects to be used, which may cause confusion and bugs later. It may also lead to duplicate calls to `Update`, if diffs sneak in after our call to `updateLabelsAndAnnotations`, which can lead to cluster-destabilizing write load on the K8s API server in heavy users of Tekton.

This may cause extra problems if Knative controller logic starts making load-bearing assumptions that it's solely responsible for the update lifecycle of objects it passes to `ReconcileKind`, which could cause headaches for us later. knative/pkg thankfully has downstream tests to warn about behavior changes that might break Tekton, but this still means they may be prevented from making otherwise helpful changes and optimizations because of how Tekton is (mis-)using its packages.

# Expected Behavior

I'd expect `ReconcileKind` not to make calls to `Update` itself, and rely on Knative's own `Update`s.

# Additional Info

- Kubernetes version: N/A, just reading code
- Tekton Pipeline version: 0.37.2

While we're at it, `updateLabelsAndAnnotations` is called by another unsuspicious-sounding method, `finishReconcileUpdateEmitEvents`, which should probably also be refactored so it's less load-bearing in the overall reconciliation cycle:

https://github.com/tektoncd/pipeline/blob/8a7b0cfa755038f4cfdcd88c314a72a90bcab1a2/pkg/reconciler/taskrun/taskrun.go#L281

@vdemeester @afrittoli WDYT?

**Comment by afrittoli** [comment]
I'm pretty sure the call to `Update` (or `Patch` before), predates cloud events. I suspect it may even predate `ReconcileKind` and perhaps it's a legacy of old times when it would actually be needed? 
We should totally try and remove it and run unit and e2e tests and see if anything breaks :)

About cloud events, moving publishing to a separate controller as pros and cons, but I think the pros definitely outweighs the cons. The main cons are:
- Some events are not matched by a dedicated status update. If a `PipelineRun` is created, and external controller could send a new "queued" event (not a started one) because it doesn't know when the main controller will pick it up from the queue and start processing it. The external controller could then send the "started" event when it detects that the start time has been set and send a  "running" event when it detects that a taskrun or run exists in the status
- Avoiding duplicate events require a local cache of events. The cache is required for `Run` events anyways, so it's in the codebase already. It was initially developed in the external cloud events controller and then ported in. The cache is in memory only at the moment, which means that a controller restart would trigger duplicate events.

The external controller exists in experimental and with @waveywaves we drafted a [roadmap](https://github.com/tektoncd/experimental/issues/830) to make it replace the current built-in support, but we then both ran out of bandwidth. I will try and pick it back up this summer.

**Comment by imjasonh** [comment]
/lifecycle frozen
/remove-lifecycle stale

I think this is still something we should do. I think @vdemeester was looking into it.

**Comment by vdemeester** [comment]
> I think this is still something we should do. I think @vdemeester was looking into it.

Yes I was looking into it recently, and it might "break" some assumption users do (on some labels being present on `PipelineRun` or `TaskRun`), so I think we might have to document and deprecate the current behavior so that we can remove it later on.

**Comment by vdemeester** [comment]
Picking this up again. After digging through the current code, here's a summary of the situation and a plan.

## Why the explicit `Update()` is still needed

The generated Knative reconciler (`pkg/client/injection/reconciler/pipeline/v1/pipelinerun/reconciler.go`) **only calls `UpdateStatus()`** after `ReconcileKind` returns — it never writes `.metadata`. So without Tekton's own `Update()`, all label/annotation mutations would be silently lost.

## What gets mutated during reconciliation

**PipelineRun:** `tekton.dev/pipeline` label (set via `propagatePipelineNameLabelToPipelineRun`), plus labels/annotations merged from the resolved Pipeline resource (in `storePipelineSpecAndMergeMeta`).

**TaskRun:** `pipeline.tekton.dev/release` annotation (set in `updateLabelsAndAnnotations` itself), plus labels/annotations merged from the resolved Task (in `storeTaskSpecAndMergeMeta`).

## The problem today

Both reconcilers call a full-object `Update()` inside `finishReconcileUpdateEmitEvents` → `updateLabelsAndAnnotations`, then Knative does an `UpdateStatus()` — two API server writes per reconcile when metadata changes. The `Update()` also bumps `resourceVersion`, which can cause the subsequent `UpdateStatus()` to conflict (retried, but wasteful under load).

Additionally, `finishReconcileUpdateEmitEvents` couples metadata persistence with event emission, making the reconciliation flow harder to follow.

## Proposed fix

1. **Replace `Update()` with a targeted `Patch(MergePatchType)`** on just `.metadata.labels` and `.metadata.annotations` — less data, no spec/status overwrite risk
2. **Decouple metadata sync from event emission** — extract it from `finishReconcileUpdateEmitEvents` into a separate `syncMetadata` call at the end of `ReconcileKind`
3. **Move `ReleaseAnnotation`** (TaskRun) to `prepare()` so it's set early like other mutations
4. **Rename `finishReconcileUpdateEmitEvents`** → `emitReconcileEvents` since it no longer does updates
5. **Keep the existing diff-check optimization** — only Patch when labels/annotations actually changed
6. **Preserve TaskRun's skip optimization** — no metadata sync when the condition was already terminal

This is fully backward-compatible — same labels/annotations end up on the objects, just via Patch instead of Update. Knative's `UpdateStatus` already uses `RetryUpdateConflicts`, so resourceVersion conflicts from the Patch are handled.

/assign
